mathphys/functions.py

```diff
@@ -197,3 +283,60 @@ def get_package_string(package):
     else:
         repo_str += f'{ver:s}'
     return repo_str
+
+
+# ------------------------- HELPER METHODS ------------------------------------
+_BUILTINTYPES = (int, float, complex, str, bytes, bool)
+_BUILTINNAMES = {typ.__name__ for typ in _BUILTINTYPES}
+_NPTYPES = (_np.int_, _np.float_, _np.complex_, _np.bool_)
+
+
+def _save_recursive_hdf5(fil, path, obj, compress):
+    typ = type(obj)
+    if isinstance(obj, _np.ndarray):
+        fil.create_dataset(path, data=obj, compression=compress)
+    elif isinstance(obj, str):  # Handle SiriusPVName
+        fil[path] = str(obj)
+    elif isinstance(obj, _BUILTINTYPES + _NPTYPES):
+        fil[path] = obj
+    elif obj is None:
+        fil[path] = 'None'
+    elif typ in {list, tuple, set}:
+        path = path.strip('/') + '/'
+        for i, el in enumerate(obj):
+            _save_recursive_hdf5(fil, path + f'{i:d}', el, compress)
+    elif typ == dict:
+        path = path.strip('/') + '/'
+        for key, item in obj.items():
+            _save_recursive_hdf5(fil, path + key, item, compress)
+    else:
+        raise TypeError('Data type not valid: '+typ.__name__+'.')
+    fil[path].attrs['type'] = typ.__name__
+
+
+def _load_recursive_hdf5(fil):
+    """."""
+    typ = fil.attrs['type']
+    if typ == 'list':
+        return [_load_recursive_hdf5(fil[str(i)]) for i in range(len(fil))]
+    elif typ == 'tuple':
+        return tuple([
+            _load_recursive_hdf5(fil[str(i)]) for i in range(len(fil))])
+    elif typ == 'set':
+        return {_load_recursive_hdf5(fil[str(i)]) for i in range(len(fil))}
+    elif typ == 'dict':
+        return {k: _load_recursive_hdf5(obj) for k, obj in fil.items()}
+    elif typ == 'NoneType':
+        return None
+    elif typ == 'ndarray':
+        return fil[()]
+    elif typ == 'str':
+        return fil[()].decode()
+    elif typ == 'SiriusPVName':
+        return fil[()].decode()
+    elif typ in _BUILTINNAMES:
+        type_ = getattr(_builtins, typ)
+        return type_(fil[()])
+    else:
+        type_ = getattr(_np, typ)
+        return type_(fil[()])
```

Comment:
Did you manage to try `return fil[()]` here?

Reply:
No, I didn't tried that. Why?

Reply:
This part of my comment from https://github.com/lnls-fac/mathphys/pull/30#discussion_r1341552991 : 

> I was testing out what h5py does, and it seems to save the type in a .dtype member of the dataset, which is different from the underlying type of the data (type seems to be optimized for storage, which I think is what happened to make the data smaller: if the max of an array fits under fewer bits, it will take advantage of it). This makes it so values stored from np.{uint,int,float,complex,bool}{8,16,32,64,128} (btw, uint is missing from _NPTYPES, unless it's a subclass of int?) are restored to the original np type if you use the [()] syntax; so I think you can skip the function call entirely, and simply return the value like you do for ndarrays. For the python int and float values, it converts to np.int64 and np.float64, respectively, so it seems to still make sense to call their respective functions.

Reply:
Sorry, @ericonr! I didn't see your comments there before resolving the issue. It didn't appear for me...

Right, I'll test it. But I think this if clause is very rare, so it won't have much effect. I'll also investigate the question regarding the types it uses to save the data. But, as we talked offline, I think the compression comes from the gzip filters, not from this automatic strategy. It would be too slow to try to check the maximum of an array before saving it. I think the syntax `[()]` is only a way to tell it to load the entire array to memory, not just a part of it.

 Maybe this is what it does to save lists or other heterogeneous collections, though, because it needs to choose a type and it's better to choose the smallest possible.

Reply:
@ericonr, since we talked offline, I think this issue is resolved, what do you think?

Reply:
I implemented as suggested.